core/vibes/soul/primitives/navigation/index.tsx

```diff
@@ -824,26 +820,43 @@ function SearchResults({
   );
 }
 
-function LocaleForm({
-  action,
+const useSwitchLocale = () => {
+  const pathname = usePathname();
+  const router = useRouter();
+  const params = useParams();
+
+  return useCallback(
+    (locale: string) =>
+      router.push(
+        // @ts-expect-error -- TypeScript will validate that only known `params`
+        // are used in combination with a given `pathname`. Since the two will
+        // always match for the current route, we can skip runtime checks.
+        { pathname, params },
+        { locale },
+      ),
+    [pathname, params, router],
+  );
```

Comment:
Use `next-intl`'s [recommended way](https://next-intl.dev/docs/routing/navigation#userouter) to switch between locales, with the exception of pushing the URL change to history (`router.push` vs `router.replace`).